### ISSUE TITLE
July iteration plan

### ISSUE THREAD

**Issue by scarlett2018** [microsoft/nni#2608]
This is the draft plan for the upcoming iteration in July, it's a 4 weeks iteration, and it is targeted to release on early August.

## Release Plan
- Release manager: @ultmaster 
- Feature freeze date: 7/31
- Code freeze and demo date: ~8/7~ 8/12
- Retrospective date: 8/14
- Release and next iteration plan date: ~8/17~ 8/24
-----
Let's try to adopt the legend of annotations:

Mark | Description
------|--------------
🏃‍♂️ | work in progress
👪 | work under review
✋ | blocked task
💪 | stretch goal for this iteration
🔴 | missing issue / pull request reference
🔵 | more investigation and survey required
⚫ | under discussion within the team
⬛ | a large work item, larger than one iteration

### Framework and SDK support
- [x] 👪 export and import experiment to make experiment be resumed or viewed on another machine #2001 @SparkSnail (ETA 7/22 for finalize design) (7/31 feature freeze) #2750 rev: @QuanluZhang @chicm-ms
  - [x] PR ready for demo
- [ ] 👪 export intermediate results and summary (both running and succeeded) nnictl change @tabVersion (Feature freeze ETA 7/24) #2706 rev: @QuanluZhang @SparkSnail pending resolving conflicts
- [X] stop experiment with port (already supported, `--port`)
- [ ] 👪 Directly display log (downloadable link) in Web UI (~needs to go through every training service and decide for each~ implement local mode only) legend PRs: #1706 #1738 PR: ~#2731~ #2718 @JunweiSUN @Lijiaoa (ETA 7/21 summary of the current to be improved things on TS) rev: @liuzhe-lz @ultmaster 
  - [x] PR ready for demo
- [x] Default log folder `/home/user/nni/experiments` -> `/home/user/nni-experiments`. @chicm-ms 7/17 rev: @ultmaster @JunweiSUN (PR #2686)
  - [ ] 👪 More fix in #2753 @JunweiSUN rev: @chicm-ms @SparkSnail 
- [x] 🔵 ~Fix evolution tuner runtime error and index error #2695~
- [x] 👪 Upgrade docker image tensorboard to 1.15 #2735 @chicm-ms rev: @ultmaster @SparkSnail 
- [x] ~Remove thop from mnist-pytorch example~
- [x] ⚫ ~Parameter id starts from 0 when experiment resumed~ @ultmaster

### Feature engineering

- [x] 👪 Fix PyTorch import issue for feature engineering @chicm-ms 7/17 rev: @xuehui1991 @ultmaster @JunweiSUN  (ETA 7/17 feature freeze) PR #2690

### NAS
- [ ] 🔵 TensorFlow v2 NAS performance issue #2781 @liuzhe-lz rev: @ultmaster @QuanluZhang (ETA 7/17 feature freeze)
- [X] 👪 Search space zoo #2589 @tabVersion rev: @ultmaster @QuanluZhang
- [X] Add intermediate result API to NAS Benchmarks #2728 @ultmaster
- [X] Fix NAS benchmarks bug #2721 #2728 @ultmaster rev: @JunweiSUN @QuanluZhang 

### Web UI
- [X] Add WebUI mock server and improve hyper-parameter parallel coordinates chart #2691 @ultmaster
- [X] 🏃‍♂️ Redesign interpolate color of hyper-parameter chart @ultmaster #2759 @Lijiaoa @colorjam 
- [ ] 🏃‍♂️ Fix issues for corner-case metrics/parameters (NaN/infinity/undefined/other-types) #2767 @Lijiaoa 
- [x] 👪 Trial details table pagination and refactor lines chart @colorjam #2738 rev: @ultmaster @Lijiaoa 
- [x] 👪 Log files drawer cannot be closed via clicking overlay area. @Lijiaoa #2734 rev: @ultmaster @liuzhe-lz 
- [x] 👪 Capitalization and spacing issues (before release). #2737 @Lijiaoa 
- [x] 🏃‍♂️ Trial No. should show sequence id instead of parameter id @JunweiSUN #2718 needs another PR @Lijiaoa #2760 rev: @squirrelsc @liuzhe-lz 
- [ ] 💪 Backlog: #2683 

### Training Service
- [X] bug: experiment cannot be resumed in pai mode @SparkSnail 
- [X] 👪 add log info for uploading codedir in training services #2702 @SparkSnail rev: @chicm-ms @ultmaster (ETA 7/24 feature freeze)
- [X] 👪 add pai job detail url in webportal #2703 @SparkSnail rev: @ultmaster @chicm-ms 
- [X] improve paimode example, change containerNFSMountPATH to /mnt/confignfs-data like paiStorageConfigName. Write placeholder in example. @SparkSnail #2685 reviewer: @squirrelsc @JunweiSUN  @scarlett2018 (ETA 7/17 feature freeze)
- [X] 👪 Training service scheduler - @squirrelsc had a PR, rev: @SparkSnail @chicm-ms  (needs review and test #2627)
  - [X] 👪 PR submitted to v1.7.1 #2692
  - [X] subtask: write tests for refactored training service framework @squirrelsc rev: @SparkSnail @QuanluZhang 
- [x] Change message in https://github.com/microsoft/nni/blob/0b9d6ce6d1cf6515c5b553a61f78cd333bf4700f/tools/nni_trial_tool/web_channel.py#L26
- [x] Set timeout for loop in trial runner @SparkSnail #2710 
- [ ] 👪 test refactoring. CI issue: test sub-directory will fail the experiment resume test @linbinskn rev: @SparkSnail @chicm-ms (ETA 7/24 feature freeze) #2694
  - [ ] by chicm-ms (8/12):  #2694 breaks windows local and remote pipelines, need to fix.
- [x] bug: remote mode issue with multi-users #2688 (PR #2689) @JunweiSUN  
- [x] 👪 support more remote servers (>20) in remote mode @SparkSnail #2725 rev: @chicm-ms @squirrelsc 
- [x] 👪 AML experiment turns into STOPPED status when azureml-sdk is not installed (needs double check) @SparkSnail #2724 rev: @JunweiSUN @ultmaster
- [x] 🔵 (by design) ~PAI job doesn't stop when experiment is stopped in reuse mode.~ @SparkSnail 

### Example, Tutorial, Jump start and doc
- [ ] 👪  nnicli intro and modify a few APIs @JunweiSUN (ETA 7/24 finalize survey) #2713 @chicm-ms @QuanluZhang 
- [x] 👪  Colab support tutorial (nnicli/frp/ngrok) #2700 @JunweiSUN rev: @ultmaster @QuanluZhang @SparkSnail 
- [x] 👪 Docker documentation (README outdated) #2720 revising #2735 @JunweiSUN  rev: @chicm-ms @SparkSnail 

### Model Compression
- [ ] 👪  TensorFlow v2 model compression support (pruner framework, one-shot algorithm) #2127 @liuzhe-lz #2755 rev: @QuanluZhang @chicm-ms 
     - [x] Review on how the current NNI Pytorch model compression implementation, survey and compare the corresponding TF 2.0 supportability. ETA 7/22
- [x] 👪 benchmark - speedup to support RetinaFace @suiguoxin #2742 rev: @chicm-ms @QuanluZhang 
  - [x] documentation #2676
- [x] 👪 Model speedup enhancement (collaborate with Yuqing team) DenseNet, Inception @zheng-ningxin #2609 #2719 rev: @QuanluZhang @chicm-ms 
- [x] 👪 Sensitivity pruner update #2684 @zheng-ningxin rev: @chicm-ms @colorjam @QuanluZhang 
- [ ] 👪 AMC (automl for model compression) paper implementation, missing docs and tests @chicm-ms rev: @QuanluZhang @colorjam  (ETA ~7/24~ 7/27 PR ready) #2573
  - [x] PR ready for demo

### 1.7.1 candidate
- [x] bug (**1.7.1 candidate**): improve reuse training service to handle earlier failure of jobs. for example, if user write incorrect path in openpai, it causes openpai job failed, but trial is waiting. The trial dispatcher will create openpai jobs, and fails continually. It needs a value like max failure to report this kind of env fatal error to nni manager, to make the whole experiment failed, and stop to create more env jobs.   (@squirrelsc had submitted a PR, reviewer: @SparkSnail) 

### Deferred

- [ ] 🏃‍♂️ Support more tuning algorithms for classic NAS (regularized evolution) @QuanluZhang #2708 rev: @ultmaster @liuzhe-lz (ETA 7/31 feature freeze) (deferred?)
  - [ ] PR ready for demo
- [ ] 🔵 NAS visualization improvements: refactor graph representation @ultmaster @zheng-ningxin (defer)
  - [ ] PR ready for demo
- [ ] 🔵 AutoML for data augmentation (image and NLP)  @chicm-ms (ETA ~7/24~ 7/31 finalize survey)
- [ ] support adding description to experiment to make users easily identify a certain experiment (defer)
- [ ] Add/Remove columns in hyper-parameter chart and trial details table and  Refactor trial details table (click event) @ultmaster 
- [ ] 👪 Cream2NNI @hwpengms rev: @ultmaster
  - [ ] PR ready for demo
- [ ] ✋ Tutorial to use tuner directly instead of using nni manager (like hyperopt?) @JunweiSUN (not started)

### Backlog
- [ ] Search Space of Tuner, quniform  +q or mode q #2397
- [ ] Fix hyperband and BOHB: support various mode, mixed multiple rounds or strictly staged rounds.
- [ ] new CV NAS algo integration @ultmaster 
- [ ] 🏃‍♂️ survey form pass internal compliance process @JunweiSUN (Pending for Jing's feedback)

**Comment by ultmaster** [comment]
Pending PR:

#2782 @liuzhe-lz @ultmaster
#2781 @ultmaster @QuanluZhang
#2780 @ultmaster @QuanluZhang @SparkSnail 
#2773 @ultmaster @Lijiaoa 
#2768 @chicm-ms 
#2755 @QuanluZhang 
#2753 Pending IT @chicm-ms 
#2747 What happens if the trial is running?
#2718 @QuanluZhang @SparkSnail @liuzhe-lz 
#2713

**Comment by ultmaster** [comment]
# Bug bash

- [ ] PAI training service 
  * Add log when copying data in PAI mode (#2702)
  * Add timeout for `web_channel` in `trial_runner` (#2710)
  * Add PAI trial job detail link (#2703)
  * Fix resume in PAI: mkdir error in training service (#2673)
- [ ] AzureML training service
  * Show environment error message in AzureML mode (#2724)
  * Support GPU scheduler in reusable environment (#2627) (#2769)
- [ ] Local training service
  * Access trial log directly on WebUI (local mode only) (#2718)
- [ ] Remote training service
  * Fix bug when using chmod in remote training service (#2689)
  * Fix support for dozens of machines in remote training service (#2725)

### WebUI, nnictl and nnicli
 
- [ ] WebUI
  * Improve hyper-parameter parallel coordinates plot (#2691) (#2759)
  * Add pagination for trial job list (#2738) (#2773)
  * Enable panel close when clicking overlay region (#2734)
  * Remove support for Multiphase on WebUI (#2760)
  * Fix several styling issues in WebUI (#2762 #2737)
  * Fix support for unusual types for metrics including NaN and Infinity (#2782)
  * Fix intermediate graph zooming issue (#2738)
- [ ] nnictl
  * Support save and restore experiments (#2750)
  * Add intermediate results in export result (#2706)
  * Add command to list trial results with highest/lowest metrics (#2747)
  * Fix nnictl experiment delete (#2791)
- [ ] nnicli
  * Improve the user experience of nnicli (#2713)
 
### Neural architecture search
 
- [ ] Search space zoo: ENAS and DARTS (#2589)
- [ ] API to query intermediate results in NAS benchmark (#2728)
  * Fix issue when dict is unordered when querying NAS benchmark (#2728)
- [ ] Fix performance issue in NAS with TensorFlow (#2781)
 
### Model compression
 
- [ ] Speedup
  * Support the List/Tuple Construct/Unpack operation for TorchModuleGraph (#2609)
  * Add support the DenseNet and InceptionV3 for the speedup module (#2719)
  * Support the multiple successive tuple unpack operations (#2768)
- [ ] Compression benchmark (#2742)
- [ ] Sensitivity pruner (#2684)
- [ ] TensorFlow v2 support in model compression (#2755)
- [ ] Support AMC pruner (#2573) (#2786)
- [ ] Fix mask issue in `SimulatedAnnealingPruner` (#2736)
- [ ] Fix dependency issue by making `_graph_utils` imported inline (#2675)

### Feature Engineering

- [ ] Fix import issue for gradient selector dataloader iterator (#2690)

**Comment by QuanluZhang** [comment]
v1.8 released
